[Woo POS] Change ItemType selection approach

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of WOOMOB-149
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR removes the notion of a `selectedItemController` for changing between showing products or coupons. Instead, the view is responsible for passing the correct ItemType when it calls functions, and for using the correct `itemState` for each list. The AggregateModel is responsible for correctly routing functions to the appropriate item controller.

This is preparation for having a third `ItemController`, for searching products. That will probably be used by a different ItemListView, but in theory they could all share one.

There are no functional changes to POS from this change – everything should work the same as before.

This is a slight inversion vs what we discussed in slack, which would have seen the itemtype wrap the base item; I found that approach was very cumbersome.

It'd be great to hear your thoughts on this approach. It feels like an improvement, but still not quite "right", in some intangiable way, to me. Still, I think it's more flexible than what we have right now.

FYI I tried cursor out to do this... obviously I've carefully reviewed and rewritten chunks of it, but as a powerful refactoring tool, it worked surprisingly well. I found that it kept putting things back in that I didn't want, as I iterated, but still it was a timesaver especially on a day where it's been hard to concentrate.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Enable the coupons in POS feature flag
2. Launch the app and open POS
3. Check that the following scenarios work properly:

- loading products, 
- loading variations, 
- loading subsequent pages,
- showing errors (full screen and inline), 
- showing the empty states
- refreshing both the above
- loading coupons
- refreshing coupons

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I noticed that the empty state isn't properly aligned... but that's pre-existing, and I'll fix it in another PR.

I also noticed that the refresh control doesn't display properly for coupons; that could be a task issue? It's pre-existing as well, and it works correctly for products, so I'm guessing it's not been implemented/polished yet.

I've tested on an iPad Air running 17.7.3, and a simulator running 18.2

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/3bf14c6f-88af-450f-af05-5808b880b000

---


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
Update item type selection approach

joshheald committed:
Tidy up minor formatting issues

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Fix lint issue; whitespace

joshheald committed:
Update tests to pass coupon type

joshheald commented:
I've noticed an issue; when on the variations list, if you change the dynamic type size the root list redraws, and it navigates back to the root list.

I think that's because we pass the node in now, or perhaps because we pass the stack. I'll investigate more tomorrow. I think this is still fine to review, with that in mind.

joshheald committed:
Update product item controller tests to pass item

staskus commented:
AI writes code:

> FYI I tried cursor out to do this... obviously I've carefully reviewed and rewritten chunks of it, but as a powerful refactoring tool, it worked surprisingly well.

AI approves code:

> Copilot reviewed 12 out of 12 changed files in this pull request and generated no comments.

joshheald committed:
DRY the determining of the ChildItemList node

staskus commented:
> I've noticed an issue; when on the variations list, if you change the dynamic type size the root list redraws, and it navigates back to the root list.

@joshheald only reproducible on iOS 17

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -172,7 +177,8 @@ private extension ItemListView {
         switch parentItem {
         case let .variableParentProduct(parentProduct):
-            ChildItemList(parentItem: parentItem, title: parentProduct.name)
+            let itemsStack = selectedItemType == .products ? posModel.itemsViewState.itemsStack : posModel.couponsViewState.itemsStack
```
Well, this is specific to items. We could use `posModel.itemsViewState.itemsStack` for `.variableParentProduct`

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift:
```diff
@@ -22,12 +22,12 @@ struct PointOfSaleDashboardView: View {
                 case .empty:
                     PointOfSaleItemListFullscreenView {
-                        PointOfSaleItemListEmptyView(base: .root)
+                        PointOfSaleItemListEmptyView(base: .root(.products))
```
Yes, correct, for coupons, we'll need to have this logic update since we wouldn't want to show a full-screen error/empty view in full-screen.

joshheald commented:
>    I've noticed an issue; when on the variations list, if you change the dynamic type size the root list redraws, and it navigates back to the root list.
> 
> @joshheald only reproducible on iOS 17


@staskus Yeah I just came to write that it's pre-existing on iOS 17 at least; it's actually due to the `floatingControlSize` environment variable changing and that redrawing the `ItemList`. I think it's caused by the iOS 17 `NavigationLink` being redrawn, as that specifies the child list destination.

I was sure I could repro it with these changes on iOS 18, but it seems I was hallucinating. Or selected the wrong simulator or something.

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -79,8 +79,9 @@ private enum Constants {
 private struct ItemListRow: View {
     let item: POSItem
-    let analytics: Analytics = ServiceLocator.analytics
+    let node: ItemListBaseItem
```
As I see `node` is only needed here to choose a correct stack. If we remove `node.itemType` check, we may be able to remove `node` from here.

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -95,18 +96,19 @@ private struct ItemListRow: View {
                 NavigationLink(destination: {
-                    ChildItemList(parentItem: item, title: parentProduct.name)
+                    let itemsStack = node.itemType == .products ? posModel.itemsViewState.itemsStack : posModel.couponsViewState.itemsStack
+                    ChildItemList(parentItem: item, title: parentProduct.name, itemsStack: itemsStack)
```
`ChildItemList` doesn't need `itemsStack` at all. We could remove the whole chain of passing `node` here.

joshheald replied:
I think we can remove node, but I think it does need the stack. It needs to know whether to use `itemsViewState` or `couponsViewState`. 

I suppose we can get away without passing for coupons/items because coupons don't have a heirarchy so will never have a child list, but the next step is to add a third itemListState for search – that will have the heirarchy, and switching on item type won't work either.

I think we can probably pass the stack in and remove the node though...

staskus replied:
Got it, thanks for the explanation.

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -36,7 +36,7 @@ struct ItemList<HeaderView: View>: View {
 
                     ForEach(state.items) { item in
-                        ItemListRow(item: item)
+                        ItemListRow(item: item, node: node)
```
```suggestion
                        ItemListRow(item: item)
```

staskus commented on WooCommerce/Classes/POS/Presentation/Item Selector/ItemList.swift:
```diff
@@ -15,7 +15,7 @@ struct ItemList<HeaderView: View>: View {
 
     init(state: ItemListState,
-         node: ItemListBaseItem = .root,
+         node: ItemListBaseItem,
```
Related to all the other comments, we may not need to be passing `node` here.

joshheald commented:
OK, it seems to happen when the `ItemListRow` which contains the navigation destination is pushed off screen in the host list. 

Since it's iOS 17 only, and will only happen for redraws which move things on the screen or remove them from the list, I think it's better off left unless I can find a solution quickly. I'm going to timebox another 30 minutes on it, then leave it as is.

staskus reviewed (commented):
Thank you!

My comments are mostly nitpicks but fixing those may allow you to view this solution in a more positive light.

> It'd be great to hear your thoughts on this approach. It feels like an improvement, but still not quite "right", in some intangiable way, to me. Still, I think it's more flexible than what we have right now.

One reason it feels not quite right because we pass `ItemListBaseItem` deep into the hierarchy but I think this could be cleared up. It should stay at a higher level, the same as the tabs are presented at a higher level. On the row level, we use `POSItem` to determine the look and behavior.

We have a flexible enough solution that even if requirements evolve, we could merge the results of coupon and item service into a single list, and the switch would only happen within `ItemListRow` where a correct view would be picked based on `POSItem`.

--

The other reason is likely the fact that we expect the inverted structure, where the item type would appear at the top level. However, if it's cumbersome to achieve then keeping `ItemType` next to the root or parent makes the most sense. It allows reusing `ItemList` in a way so it could be supported by all types of base or parent items and request correct list reload strategy. I tried to play with some other approaches but they failed one way or another.

staskus reviewed (approved):
> Since it's iOS 17 only, and will only happen for redraws which move things on the screen or remove them from the list, I think it's better off left unless I can find a solution quickly. I'm going to timebox another 30 minutes on it, then leave it as is.
 
 Okay. That seems reasonable. I'm then approving it with some suggested changes. I spent quite a lot of time on this review playing around with structure. I agree that it _feels_ like it could be a bit different but I couldn't quite find the right balance myself.

joshheald commented:
Thanks for the time and comments @staskus

joshheald committed:
Determine itemstack outside the ItemList

This will allow us to have two itemlists of the same type (e.g. search) rely on a single posModel, without having to muddy the waters of the ItemList with code to check what list to get data from.

joshheald committed:
Keep list state when font size changes on iOS17

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Fix whitespace

joshheald committed:
Merge branch 'trunk' into woomob-149-change-itemtype-selection-approach
